File: src/services/cherryPick.ts
import { invoke } from "@tauri-apps/api/core";

export type CherryPickMode = "normal" | "no_commit" | "merge";

export interface CherryPickResult {
  completed: boolean;
  conflicts: string[];
  oid: string | null;
}

export function cherryPick(
  tabId: string,
  oids: string[],
  mode: CherryPickMode,
): Promise<CherryPickResult> {
  return invoke<CherryPickResult>("cherry_pick", { tabId, oids, mode });
}

export function isCherryPicking(tabId: string): Promise<boolean> {
  return invoke<boolean>("is_cherry_picking", { tabId });
}

export function abortCherryPick(tabId: string): Promise<void> {
  return invoke<void>("abort_cherry_pick", { tabId });
}

export function continueCherryPick(tabId: string): Promise<CherryPickResult> {
  return invoke<CherryPickResult>("continue_cherry_pick", { tabId });
}

File: src-tauri/src/commands/cherry_pick.rs
use tauri::State;

use crate::commands::with_repo;
use crate::git::types::{CherryPickMode, CherryPickResult};
use crate::state::AppState;

#[tauri::command]
pub fn cherry_pick(
    tab_id: String,
    oids: Vec<String>,
    mode: CherryPickMode,
    state: State<'_, AppState>,
) -> Result<CherryPickResult, String> {
    with_repo(&state, &tab_id, |backend| {
        let oid_refs: Vec<&str> = oids.iter().map(|s| s.as_str()).collect();
        backend
            .cherry_pick(&oid_refs, mode)
            .map_err(|e| e.to_string())
    })
}

#[tauri::command]
pub fn is_cherry_picking(tab_id: String, state: State<'_, AppState>) -> Result<bool, String> {
    with_repo(&state, &tab_id, |backend| {
        backend.is_cherry_picking().map_err(|e| e.to_string())
    })
}

#[tauri::command]
pub fn abort_cherry_pick(tab_id: String, state: State<'_, AppState>) -> Result<(), String> {
    with_repo(&state, &tab_id, |backend| {
        backend.abort_cherry_pick().map_err(|e| e.to_string())
    })
}

#[tauri::command]
pub fn continue_cherry_pick(
    tab_id: String,
    state: State<'_, AppState>,
) -> Result<CherryPickResult, String> {
    with_repo(&state, &tab_id, |backend| {
        backend.continue_cherry_pick().map_err(|e| e.to_string())
    })
}
